src/Agent.Worker/Build/TfsVCTeeSourceProvider.cs

```diff
@@ -0,0 +1,373 @@
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Agent;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Agent.Worker;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Build
+{
+    public sealed class TfsVCTeeSourceProvider : SourceProvider, ISourceProvider
+    {
+        private string _workspaceName;
+
+        public override string RepositoryType => WellKnownRepositoryTypes.TfsVersionControl;
+
+        public async Task GetSourceAsync(
+            IExecutionContext executionContext,
+            ServiceEndpoint endpoint,
+            TrackingConfig trackingConfig,
+            CancellationToken cancellationToken)
+        {
+            // Validate args.
+            ArgUtil.NotNull(executionContext, nameof(executionContext));
+            ArgUtil.NotNull(endpoint, nameof(endpoint));
+            ArgUtil.NotNull(trackingConfig, nameof(trackingConfig));
+
+            // Create the tf command manager.
+            var tf = HostContext.CreateService<ITeeTFCommandManager>();
+            tf.CancellationToken = cancellationToken;
+            tf.Endpoint = endpoint;
+            tf.ExecutionContext = executionContext;
+
+            // Get the TEE workspaces.
+            TeeWorkspace[] teeWorkspaces = await tf.WorkspacesAsync();
+
+            // Determine the workspace name.
+            AgentSettings settings = HostContext.GetService<IConfigurationStore>().GetSettings();
+            _workspaceName = $"ws_{trackingConfig.BuildDirectory}_{settings.AgentId}";
+
+            // Get the definition mappings.
+            TfsVCTeeWorkspaceMapping[] definitionMappings =
+                JsonConvert.DeserializeObject<TfsVCTeeWorkspaceMappings>(endpoint.Data["tfvcWorkspaceMapping"])?.Mappings;
+
+            // Determine the sources directory.
+            string sourcesDirectory = Path.Combine(IOUtil.GetWorkPath(HostContext), trackingConfig.SourcesDirectory);
+
+            // Attempt to re-use an existing workspace if clean=false.
+            TeeWorkspace existingTeeWorkspace = null;
+            bool clean = endpoint.Data.ContainsKey(WellKnownEndpointData.Clean) &&
+                StringUtil.ConvertToBoolean(endpoint.Data[WellKnownEndpointData.Clean], defaultValue: false);
+            if (!clean)
```

Comment:
I played with the available tf commands *forever*. It indeed doesn't appear that scorch is possible.